docs/design/federated-service-discovery.md

```diff
@@ -0,0 +1,403 @@
+preferentially connect to the shard of the service running within the same
+cluster, if one exists. Otherwise, they should be able to connect to the shard
+of the service within the same zone or the same region in that order of
+preference. If no such shards exist, then they should be able to connect to any
+shard of the service. In addition, clients should be able to explicitly specify
+the shard of the service they want to connect to.
+
+# Background
+
+Kubernetes provides service discovery through two different mechanisms:
+
+  1. Environment variables
+  2. DNS
+
+In Ubernetes, we do not intend to support service discovery through environment
+variables for various reasons discussed in the [Federated (Ubernetes) Service Discovery DNS details - rough notes](https://docs.google.com/a/google.com/document/d/1Lqe8rhiLveoDRoRQHRqL5p91_yQ1M9wIT8THvbTVH4M/edit?usp=sharing)
+doc. We will continue to support the mechanisms provided by Kubernetes for the
+local shards of federated services, just as if they were services within a
+single Kubernetes cluster, but clients will not be able to discover shards in
+other clusters using this mechanism.
+
+Service discovery for federated services will be provided through DNS. [Federated (Ubernetes) Service Discovery DNS details - rough notes](https://docs.google.com/a/google.com/document/d/1Lqe8rhiLveoDRoRQHRqL5p91_yQ1M9wIT8THvbTVH4M/edit?usp=sharing)
+and [Federated Services](federated-services.md)
+docs have more detailed discussions on the requirements of such a DNS-based
+discovery scheme. Please refer to those docs for details and additional
+background.
+
+# Design
+
+The overall architecture of the service discovery mechanism for federated
+services in Ubernetes is shown in the diagram below.
+
+<!-- TODO: Update the rawgit.com URL to point to kubernetes/kubernetes before merging this PR. -->
+![Federated Service Discovery Architecture](https://rawgit.com/madhusudancs/kubernetes/federated-service-discovery/docs/design/federated-service-discovery-arch.svg)
+
+## Apparatus
+
+Each individual Kubernetes cluster runs its own DNS server, an instance of
+`SkyDNS`, as a cluster addon that provides service discovery for cluster-local
+services. Every container in every pod in the cluster is configured to use this
+DNS server instance for domain name resolution. The rest of this doc refers to
+this DNS server instance as **`K8S-DNS`**.
+
+**`K8S-DNS`** runs an instance of `SkyDNS` along with its own `etcd` and
+`kube2sky` controller in a Kubernetes pod. Each of these components run in
+their own container within the pod. `kube2sky` watches the services in the
+cluster and programs the etcd backing the `SkyDNS`. In addition to these
+components, another container running an HTTP server, called stubzone server,
+is introduced into this pod to facilitate Ubernetes service discovery. This
+HTTP server exposes a single endpoint that allows programming the
```

Comment:
ah, I see the TLS client cert a little later.  I'd add that to the diagram.  Maybe a lock or something.